### DISCUSSION TITLE
Standalone LSP clients?

### DISCUSSION THREAD

**Discussion by damnskippy** [clangd/clangd#978] [Q&A]
Are there any standalone LSP client implementations that are not attached to a specific editor, but take inputs and operate from the command line?
Thanks.

**Comment by sam-mccall** [comment]
Interesting idea, not as far as I can tell.
A couple of concerns:
 - the obvious usability one that most operations use a location as input, and passing coordinates on the command-line is annoying
 - performance: clangd (and at least some other language servers) are pretty reliant on a stateful cache. They take a relatively long time to work at all, and in some cases give degraded functionality until (e.g.) an index is prepared. Starting a language server, performing one operation, and then shutting it down wouldn't be ideal.

**Reply by damnskippy** [reply]
Completely agreed, I was thinking more along the lines of server being up standalone, and client is feeding inputs and getting results back, similar to what the editor environment would do.  But perhaps this is more than likely the reason why there isn't a standalone client.  The use case is probably not quite there as much as in a editor.  Wonder if there are things that are possible to derive useful insights via LSP if such a standalone client were to exist, perhaps retrieve a call graph for example via incoming/outgoing call hierarchy, or any class dependencies, etc.  Not that any of this is not possible to do via an editor already.  Thanks for capturing the reasons about expensive startup/shutdown/caching as why this is not an ideal approach.

**Comment by HighCommander4** [comment]
May not be exactly what you're looking for, but I've found https://github.com/simark/ls-interact useful in the past for testing and scripting LSP interactions

**Reply by damnskippy** [reply]
Yes, most definitely along the lines of what I was thinking.  Interesting.  Thanks.

**Reply by Hebe970217** [reply]
This is an interesting idea.
I'm cpp developer. 
If you want to better understand the protocol, you should find a publishing code editing tool familiar with the language. 
For example, I found ide "Kate"

**Comment by i-ky** [comment]
I would suggest checking out the ["official" list of SDKs implementing LSP](https://microsoft.github.io/language-server-protocol/implementors/sdks/), choosing one in the language you feel comfortable with and writing a small wrapper script. I had some positive experience with [pylspclient](https://github.com/yeger00/pylspclient). Even though it does not provide "pythonic" interface for all LSP features, it was still easy to hack around and send custom JSON requests based on LSP specification.
